File: C++/15.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        for(int i = 0; i < nums.size()-2;) {
            int k = nums.size()-1;
            for(int j = i+1; j < nums.size()-1;) {
                while(k > j && nums[i] + nums[j] + nums[k] > 0) k--;
                if(k <= j) break;
                if(nums[i] + nums[j] + nums[k] == 0) {
                    result.push_back(vector<int>({nums[i], nums[j], nums[k]}));
                    k--;
                }
                j++;
                while(j < nums.size()-1 && nums[j] == nums[j-1]) j++;
            }
            i++;
            while(i < nums.size()-2&&nums[i] == nums[i-1]) i++;
        }
        return result;
    }
};

File: Python/15.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        n = len(nums)
        res = []
        for i in range(n):
            p = i+1
            q = n-1
            while p < q:
                if nums[p] + nums[q] == -nums[i]:
                    res.append([nums[i], nums[p], nums[q]])
                    p += 1
                    q -= 1
                elif nums[p] + nums[q] > -nums[i]:
                    q -= 1
                else:
                    p += 1
        res=list(set([tuple(t) for t in res]))
        return res